### ISSUE TITLE
N64 Style Depth Compare graphical issues on NVidia

### ISSUE THREAD

**Issue by quattj** [gonetz/GLideN64#2328]
Windows 7 64 bit, current video card MSI RTX 2060 Super Gaming X, NVidia driver 452.06
Previous video cards: XFX Radeon 460/560 and Sapphire Radeon 5600 XT using latest Radeon software as of one month ago
Emulator: m64p

On my previous Radeon cards I was able to use N64 Style Depth Compare with no issues on "Fast".  All my games ran at 30 FPS with no issues.
With my new NVidia card, "Fast" still works at the same speed, but all of my games have graphical artifacts in the textures, with random small squares in the 3D textures flickering, twitching, etc (2D items do not appear to be affected).  If I switch it to "Compatible", the graphical issues disappear, but then certain games get slowdown (Mario Kart 64, all races stutter on start then smooth out in multi-player, certain races drop down to 20 FPS at certain points).

I just updated to the latest m64p from my previous version from earlier this year with no improvement in the issues.

![mariokart64-000](https://user-images.githubusercontent.com/63088660/93044747-7f678380-f623-11ea-98b2-31426cf4ec3b.png)
![wheel_of_fortune-006](https://user-images.githubusercontent.com/63088660/93044981-116f8c00-f624-11ea-9875-ff17b8e68a5a.png)
![wheel_of_fortune-007](https://user-images.githubusercontent.com/63088660/93044982-12082280-f624-11ea-91c0-3aef03503b93.png)
![mariokart64-000](https://user-images.githubusercontent.com/63088660/93044983-12a0b900-f624-11ea-8e23-76ae8f28b488.png)
![wheel_of_fortune-000](https://user-images.githubusercontent.com/63088660/93044986-13394f80-f624-11ea-9209-3c8bf6240094.png)
![wheel_of_fortune-001](https://user-images.githubusercontent.com/63088660/93044987-13394f80-f624-11ea-9fb5-4af29006eab1.png)

**Comment by weinerschnitzel** [comment]
is this the same as #2256? I wonder if the approach suggested in #1970 would work better for all cards.

**Comment by quattj** [comment]
Yes, it appears to be the same as #2256 It is unfortunate because many games seem to have more accurate rendering with it enabled.

**Comment by gonetz** [comment]
> I wonder if the approach suggested in #1970

Which approach was suggested? per-pixel linked lists ?

**Comment by weinerschnitzel** [comment]
Yes, I was curious about the per-pixel linked lists proposal

**Comment by oddMLan** [comment]
I heard that's how other emulators do it (some Dreamcast emulator)? But the requirements increase exponentially with a higher internal resolution. Curse you AMD!

**Comment by CadetSparklez** [comment]
> 
> 
> Yes, it appears to be the same as #2256 It is unfortunate because many games seem to have more accurate rendering with it enabled.

what have you noticed it more accurate in? I tried it in PD but didnt notice a difference, so I shut it off to get my msaa back

**Comment by quattj** [comment]
> 
> 
> > Yes, it appears to be the same as #2256 It is unfortunate because many games seem to have more accurate rendering with it enabled.
> 
> what have you noticed it more accurate in? I tried it in PD but didnt notice a difference, so I shut it off to get my msaa back

Just in general.  I can't pinpoint much off the top of my head, but small things like the near-to-screen cutoff in Mario Kart.  Without it the textures are cut a little further away from the screen, giving more rainbow-y effects when zoomed out during a race and driving through certain areas where the edges of the track are elevated and are close to the screen.  I think a few spots in Jeopardy are affected.  May be others, but I haven't checked too many lately as I've been playing games for *gasp* other systems ;)  I use the force widescreen rendering option for most of my N64 games.  It's not a major issue, just slight differences that you could probably tell if they were side-by-side.

**Comment by CallistoNTG** [comment]
> > Yes, it appears to be the same as #2256 It is unfortunate because many games seem to have more accurate rendering with it enabled.
> 
> what have you noticed it more accurate in? I tried it in PD but didnt notice a difference, so I shut it off to get my msaa back

For a start, none of the decals render correctly in Perfect Dark without it. Blood splatter, bullet holes, burn makes, etc are not clipped properly along edges. And the cloaking effects are wrong, especially on the Mauler. This is the same reason why shadows are wrong in Banjo Tooie without it.

**Comment by CadetSparklez** [comment]
> 
> For a start, none of the decals render correctly in Perfect Dark without it. Blood splatter, bullet holes, burn marks, etc are not clipped properly along edges. And the cloaking effects are wrong, especially on the Mauler. This is the same reason why shadows are wrong in Banjo Tooie without it.

oh yeah I know what you are talking about with the mauler, guess I never noticed the decals. I always have paintballs on instead of blood and those only show up on flat surfaces not bodies so I never really noticed. As for bullet holes, I didnt even remember those existed since on the n64 I would never notice unless I went right up to the wall.
thx for the help. ill prob keep msaa on instead of some minor clipping but its good to know what that did!

**Comment by GhostlyDark** [comment]
Easy to spot running either Zelda game. 3000 series with 5xx drivers, oldest compatible driver or newest one. It doesn't matter.
Unless the card is extremly powerful (3090 level of performance?), it won't be able to run with the `Compatible` setting at full speed. 1080 Ti has slowdowns for instance.

![image](https://user-images.githubusercontent.com/31742919/184549812-858b985a-e221-48f7-bded-b07b116257aa.png)

**Comment by Xii-Nyth** [comment]
> For a start, none of the decals render correctly in Perfect Dark without it. Blood splatter, bullet holes, burn marks, etc are not clipped properly along edges. And the cloaking effects are wrong, especially on the Mauler. This is the same reason why shadows are wrong in Banjo Tooie without it.

Today is it just the surface hit sparks that have issues? Or were those always buggy even on the n64?
I sometimes don't see them appear when they should, but I was guessing its just the surface type being set wrong in the game.

They are a little odd since in goldeneye they showed up as a texture with omnidirectional sparks, but in pd they are drawn directly from code like the radar dots and and starts, or at least done indirectly, as I can't dump their texture. I'm guessing it was done like this so they can appear based off of the angle the surface what hit rather than in all directions but it doesn't seem to have any consistency to it.
